File: net/Triangle.cpp

using namespace Engine;

Shapes::Triangle::Triangle(const Vector a, const Vector b, const Vector c, IMaterial material): 
	a(a),
	ba(b - a),
	ca(c - a),
	normal((b - a).CrossProduct(c - a)),
	n((b - a).CrossProduct(c - a).Normalize()),
	material(material)
{
}

HitPoint Shapes::Triangle::Intersection(Vector start, Vector direction) const
{
	double t = 0;
	double t1 = 0;
	double t2 = 0;
	

	double divident = - direction.DotProduct(normal);
	
	if(!divident)
	{
		return NULL;
	}

	double factor = 1 / divident;

	Vector sa = start - a;
	Vector saxdir = sa.CrossProduct(direction);
	t1 = -ba.DotProduct(saxdir) * factor;
	
	if((t1 < 0) || (1 < t1))
	{
		return NULL;
	}
				
	t2 = ca.DotProduct(saxdir) * factor;
	
	if((t2 < 0) || (1 < t2))
	{
		return NULL;
	}
	
	if((t2 + t1) > 1)
	{
		return NULL;
	}
	
	t = sa.DotProduct(normal) * factor;

	if(t < double_EPSILON)
	{
		return NULL;
	}

	return new HitPoint(t, n, material);
}

File: net/Shapes/Triangle.cs
using Engine;

namespace Shapes;

public class Triangle: IShape
{
	public readonly Vector baCrossCa;
	public readonly Vector normal;
	public readonly Vector ba;
	public readonly Vector ca;
	public readonly Vector a;

	public readonly AABB aabb;
	public readonly bool oneSide;

	public Triangle(Vector a, Vector b, Vector c, bool oneSide = false)
	{
		this.a = a;
		this.ba = b - a;
		this.ca = c - a;
		this.baCrossCa = ba.CrossProduct(ca);
		this.normal = ba.CrossProduct(ca).Normalized;
		this.oneSide = oneSide;
		aabb = AABB.FromEdgePoints(
			a,
			b, 
			c 
		);
	}

	public ref readonly AABB AABB => ref aabb;

	public ShapeHitPoint Intersection(in Ray ray)
	{
		var divident = -ray.direction.DotProduct(baCrossCa);
		
		if(divident == 0 || oneSide && divident < 0)
		{
			return null;
		}

		var factor = 1 / divident;

		var sa = ray.start - a;
		var saCrossDir = sa.CrossProduct(ray.direction);
		var t1 = -ba.DotProduct(saCrossDir) * factor;
		
		if((t1 < 0) || (1 < t1))
		{
			return null;
		}
					
		var t2 = ca.DotProduct(saCrossDir) * factor;
		
		if((t2 < 0) || (1 < t2))
		{
			return null;
		}
		
		if((t2 + t1) > 1)
		{
			return null;
		}
		
		var t = sa.DotProduct(baCrossCa) * factor;

		if(t < 0)
		{
			return null;
		}

		return new ShapeHitPoint(ray, t, oneSide ? normal : (normal * Math.Sign(divident)));
	}
}
